keras_hub/src/models/deepseek_r1/deepseek_layers.py

```diff
@@ -0,0 +1,786 @@
+import math
+
+import torch
+from keras import layers
+from keras import ops
+
+from keras_hub.src.layers.modeling.rms_normalization import RMSNormalization
+
+attn_impl = "absorb"
+rank = 0
+
+
+class Embedding(layers.Layer):
+    """
+    Embedding layer with parallelism support across distributed processes.
+
+    Args:
+        vocab_size (int): Vocabulary size.
+        dim (int): Embedding dimension.
+    """
+
+    def __init__(self, vocab_size: int, dim: int):
+        super().__init__()
+        self.dim = dim
+        self.vocab_size = vocab_size
+        self.embedding = layers.Embedding(self.vocab_size, self.dim)
+
+    def call(self, x):
+        return self.embedding(x)
+
+
+# This can probably be replaced by a Dense layer.
+# Does it transpose the weight?
+def linear(x, weight, bias=None):
+    """
+    Applies a linear transformation to the incoming data: y = xA^T + b.
+
+    Args:
+        x: The input tensor.
+        weight: The weight tensor.
+        bias: The bias tensor to be added. Default is None.
+
+    Returns:
+        The result of the linear transformation.
+
+    """
+    x = ops.matmul(x, ops.transpose(weight))
+    if bias:
+        x = x + bias
+
+    return x
+
+
+class Linear(layers.Layer):
+    """
+    Custom linear layer with support for optional bias.
+
+    Args:
+        in_features (int): Number of input features.
+        out_features (int): Number of output features.
+        bias (bool): Whether to include a bias term. Defaults to False.
+        dtype (optional): Data type for the layer.
+    """
+
+    def __init__(self, in_features, out_features, bias=False):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+
+        self.weight = self.add_weight(
+            shape=(out_features, in_features), trainable=True, name="weight"
+        )
+
+        if bias:
+            self.bias = self.add_weight(
+                shape=(out_features,), trainable=True, name="bias"
+            )
+        else:
+            self.bias = None
+
+        # Scale was None if weight's element size was != 1
+        # It's only =1 in the case of uint8.
+        scale_out_features = (out_features + block_size - 1) // block_size
+        scale_in_features = (in_features + block_size - 1) // block_size
+        self.weight.scale = self.scale = nn.Parameter(
+            torch.empty(
+                scale_out_features, scale_in_features, dtype=torch.float32
+            )
+        )
+
+    def call(self, x):
+        return linear(x, self.weight, self.bias)
+
+
+class ColumnParallelLinear(Linear):
```

Comment:
Do we still need the custom `XParallel` classes if we don't use torch.dist which boils most of them back to the standard implementations?